src/dinitctl.cc

```diff
@@ -1903,39 +1903,61 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
     }
 
     if (socknum >= 0) {
+        if (verbose) {
+            cout << "Service '" << to << "' has been " << (enable ? "enabled" : "disabled") << "." << endl;
+        }
+
         // Check status of the service now
         auto m = membuf()
                 .append((char)cp_cmd::SERVICESTATUS)
                 .append(to_handle);
         write_all_x(socknum, m);
 
-        wait_for_reply(rbuffer, socknum);
-        if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {
+        int r = rbuffer.fill_to(socknum, 2);
+        while (r > 0 && rbuffer[0] >= 100) {
+            unsigned pktlen = (unsigned char) rbuffer[1];
+            fill_buffer_to(rbuffer, socknum, pktlen);
+
+            if (rbuffer[0] == (char)cp_info::SERVICEEVENT) {
+                wait_service_state(socknum, rbuffer, to_handle, to, !enable, verbose);
```

Comment:
Doing this for a _disable_ changes the operation. A _disable_ didn't use to wait for the service state to change, and that was intentional. (We only wait for it on _enable_ because if a user enables a service they probably want to know if it fails to start).

Also, just because a `SERVICEEVENT` was received doesn't mean that `wait_service_state` will only process a single event packet nor that it will (ever) see one of the events it is looking for. For _enable_, `wait_service_state` will wait for either `STARTED` or `STARTCANCELLED` but there is no guarantee that either will arrive before the `SERVICESTATUS` reply, and `wait_service_state` will throw a protocol error if that arrives, because is not designed to be used while waiting for a reply packet. So, you can't use it here.

Eg consider:

- ENABLESERVICE sent by dinitctl
- service is already started, ACK is queued in daemon
- service stops, SERVICEEVENT queued (with STOPPED status)
- dinitctl receives ACK
- SERVICESTATUS request sent by dinitctl
- SERVICEEVENT received by dinitctl - calls `wait_service_state` 
- SERVICESTATUS reply received, `wait_service_state` throws protocol error

There was no actual protocol error, the error was that `wait_service_state` was used when a reply packet was pending.

Reply:
The new `process_service_event()` is here to just process a service event rather than waiting for service events (unlike the `wait_service_state()`). now `wait_service_state()` and `enable_disable_service()` uses this new function.